[Hack Week] ApiFaker: Adding, deleting and deleting endpoints

hichamboushaba opened the pull request:
### Description
As part of my Hack Week, I'm bringing back the ApiFaker project paqN3M-SI-p2, and I'm splitting the feature into multiple PRs.

This PR adds the endpoint details screen, which is used to add or edit the endpoint details, it also adds logic for deleting endpoints.

**Sorry for the size of the PR, but I couldn't split it further.**

### Steps to reproduce
1. Use a debug build.
2. Open the settings page.
3. Tap on Developer Options.
4. Tap on ApiFaker.

### Testing information
- Test adding, editing, and deleting endpoints.
- Test endpoint faking logic (preferably using both WordPress.com and site credentials logins)

### The tests that have been performed
^

### Images/gif
<img src=https://github.com/user-attachments/assets/31e5762a-0c00-466e-aa98-319a3c63937f width=360 />
<img src=https://github.com/user-attachments/assets/82e935a7-c150-42b5-a20f-e23555c67a6c width=360 />

</br>
</br>

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Start working on the details screen

hichamboushaba committed:
Handle remaining fields

hichamboushaba committed:
Handle saving changes

hichamboushaba committed:
Fixes to saving and navigation

hichamboushaba committed:
Rename classes

hichamboushaba committed:
Improve the query statement

hichamboushaba committed:
Minor improvements

hichamboushaba committed:
Consume the status configuration

hichamboushaba committed:
Add ability to delete requests

hichamboushaba committed:
Add support for filtering by http method

hichamboushaba committed:
Display a hint when ApiFaker is enabled

hichamboushaba committed:
Improve navigation

hichamboushaba committed:
UI improvements

hichamboushaba committed:
Apply internal visibility to some declarations

hichamboushaba committed:
Add ability to disable the ApiFaker from the UI hint

hichamboushaba committed:
Add query parameters to the DB model

hichamboushaba committed:
Filter requests using query parameters

hichamboushaba committed:
Add a way to add query parameters in the UI

hichamboushaba committed:
Some fixes

hichamboushaba committed:
Log a warning message if multiple endpoints are matched

hichamboushaba committed:
Add an artifical delay

hichamboushaba committed:
Suppress deprecation temporarily

dangermattic commented:
<!--
  3 Errors: Please add tests for class <co...
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          3 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>ApiFakerUiHelper</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>QueryParameterConverter</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>EndpointDetailsViewModel</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Bump db version

hichamboushaba committed:
Handle query parameters extraction for Jetpack tunnel requests

hichamboushaba committed:
Set text color as white in the hint

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Add some padding between the text fields and their captions

hichamboushaba commented on libs/apifaker/src/main/java/com/woocommerce/android/apifaker/ui/home/HomeScreen.kt:
```diff
@@ -109,47 +125,81 @@ private fun HomeScreen(
+        state = dismissState,
+        directions = setOf(DismissDirection.EndToStart),
+        dismissThresholds = {
+            @Suppress("DEPRECATION")
+            androidx.compose.material.FractionalThreshold(0.3f)
```
I'm temporarily suppressing the deprecation here as I'm just porting the code from the old branch.
I'll have a PR that migrates this to the new API after this.

hichamboushaba committed:
Add more tests

JorgeMucientes reviewed (approved):
Well done @hichamboushaba the code looks good and so does the UI! 

I just found a minor glitch on the UI. When the endpoint is long enough, the error code won't be displayed in the saved endpoints list screen:

![Screenshot 2024-12-11 at 10 11 37](https://github.com/user-attachments/assets/7e5a3efa-d373-493c-8477-1f2281320fb9)

Aside from that, would it be possible to add some instructions to intercept Jetpack tunnel requests? It took me a while to figure out it was this combination that would make it work with the jetpack tunnel. I started adding it as a WP.com REST API since jetpack requests have `https://public-api.wordpress.com/rest/` as initial path. Then I realized I had to set it as WordPress REST api and only add the path after the `path=` query param.  In fact I had to debug the code to figure out what API faker expects to match the request. Maybe it was just me, but it didn't feel intuitive as with WP.com REST API.

hichamboushaba commented:
Thanks @JorgeMucientes for the review.

> I just found a minor glitch on the UI. When the endpoint is long enough, the error code won't be displayed in the saved endpoints list screen:

Nice catch, I'll fix it before merging the PR.

> would it be possible to add some instructions to intercept Jetpack tunnel requests? It took me a while to figure out it was this combination that would make it work with the jetpack tunnel. I started adding it as a WP.com REST API since jetpack requests have https://public-api.wordpress.com/rest/ as initial path. Then I realized I had to set it as WordPress REST api and only add the path after the path= query param. In fact I had to debug the code to figure out what API faker expects to match the request. Maybe it was just me, but it didn't feel intuitive as with WP.com REST API.

Yes, this is a tricky point about our setup. My goal here is for the user (developer) to avoid thinking about Jetpack tunnel as a whole, just like with our networking layer in the app, when we want to call `/wc/v3/products`, we just pass this path to the `WooNetwork`, and it would use the correct full URL depending on the setup, the same happens with the `ApiFaker`, we give it just the `path`, and it can fake the request for both site credentials and WordPress.com (Jetpack tunnel).
I'll think of a way to make this clearer in the UI, but I'll do it in a separate PR if you don't mind.

hichamboushaba committed:
Fix UI with long endpoint paths

hichamboushaba committed:
Improve logic of warning log

This makes sure the log is not printed when the matched endpoints use different query parameters.

hichamboushaba committed:
Merge pull request #13083 from woocommerce/task/api-faker-5

[Hack Week] ApiFaker: match endpoints using query parameters

hichamboushaba committed:
Merge pull request #13076 from woocommerce/task/api-faker-4

[Hack Week] ApiFaker: display a hint on the main app UI when the ApiFaker is enabled
